aws/session/session.go

```diff
@@ -21,36 +15,128 @@ import (
 // store configurations and request handlers for those services.
 //
 // Sessions are safe to create service clients concurrently, but it is not safe
-// to mutate the session concurrently.
+// to mutate the Session concurrently.
+//
+// The Session satisfies the service client's client.ClientConfigProvider.
 type Session struct {
 	Config   *aws.Config
 	Handlers request.Handlers
 }
 
-// New creates a new instance of the handlers merging in the provided Configs
-// on top of the SDK's default configurations. Once the session is created it
-// can be mutated to modify Configs or Handlers. The session is safe to be read
+// New creates a new instance of the handlers merging in the provided configs
+// on top of the SDK's default configurations. Once the Session is created it
+// can be mutated to modify the Config or Handlers. The Session is safe to be
+// read concurrently, but it should not be written to concurrently.
+//
+// If the AWS_SDK_LOAD_CONFIG environment is set to a truthy value, the New
+// method could now encounter an error when loading the configuration. When
+// the environment variable is set New will panic if an error occurs. The Build
+// methods should be used instead to avoid this panic, as they will return an
+// error if one occurs.
+//
+// If the AWS_SDK_LOAD_CONFIG environment variable is set to a truthy value
+// the shared config file (~/.aws/config) will also be loaded, in addition to
+// the shared credentials file (~/.aws/config). Values set in both the shared
+// shared config, and shared credentials will be taken from the shared
+// credentials file.
+//
+// Deprecated: Use Build instead. Build has the same functionality as New
+// except an error is returned instead of panic.
+func New(cfgs ...*aws.Config) *Session {
+	// load initial config from environment
+	envCfg := loadEnvConfig()
+
+	if envCfg.EnableSharedConfig {
+		// Old session.New expected all errors to be discovered when
+		// a request is made, and would report the errors then. This
+		// needs to be mocked.
+		s, err := buildSession(envCfg, cfgs...)
+		if err != nil {
+			// TODO error message
+			panic(err)
+		}
+		return s
+	}
+
+	return oldNewSession(cfgs...)
+}
+
+// Must is a helper function to ensure the Session is valid and there was no
+// error when calling a Build function.
+//
+// This helper is intended to be used in variable initialization to load the
+// Session and configuration at startup. Such as:
+//
+//     var sess = session.Must(session.Build())
+func Must(sess *Session, err error) *Session {
```

Comment:
That sounds fine, can remove this method. It has a nitch usage and can add it back later if there is demand for it.